### PR TITLE
Census Keras input generator fix

### PR DESCRIPTION
**Pull request by RemiTurpaud**
Fix to enable the input generator to iterate over multiple file chunks and return to the file begging when reaching the end.
Also updated the feature extractor `to_numeric_features()` to return a consistent dataframe schema: the categorical classes are 'one hot' encoded based on the available values in the processed chunk, hence the returned column list may vary from a chunk to another (if new values are found or old are missing).

Added a mini-batch functionality to the generator_input (previous version returning one sample at a time). Training on generated mini-batches considerably* improves the performances. (I do not know how to reliably measure this on Google Cloud - this version seems to consume a 5th of the ML units and a 20th of the time, compared to the previous one, for the same processed volume). 

Finally, changed the model defaults to better fit the problem (and show some learning with the default example):
- softmax for the final activation
- RMSprop for the optimizer (along with the default learning rate)

### CODE REVIEW

File: census/keras/trainer/model.py [hunk 1]
```diff
@@ -16,14 +16,12 @@
...
 """Implements the Keras Sequential model."""
 
-import itertools
-
 import keras
 import pandas as pd
 from keras import backend as K
 from keras import layers, models
 from keras.utils import np_utils
-from keras.backend import relu, sigmoid
```
Comment (puneith) [inline census/keras/trainer/model.py]: @RemiTurpaud Why is this line replaced with sigmoid? Its not a 2-class problem.

File: census/keras/trainer/model.py [hunk 2]
```diff
@@ -69,13 +67,13 @@ def model_fn(input_dim,
...
   # Add a dense final layer with sigmoid function
-  model.add(layers.Dense(labels_dim, activation=sigmoid))
+  model.add(layers.Dense(labels_dim, activation=softmax))
   compile_model(model, learning_rate)
   return model
 
 def compile_model(model, learning_rate):
   model.compile(loss='categorical_crossentropy',
-                optimizer=keras.optimizers.SGD(lr=learning_rate),
+                optimizer=keras.optimizers.RMSprop(lr=learning_rate),
```
Comment (puneith) [inline census/keras/trainer/model.py]: Is RMSProp giving better result in this case?

File: census/keras/trainer/model.py [hunk 3]
```diff
@@ -16,14 +16,12 @@
...
 
-import itertools
-
 import keras
 import pandas as pd
 from keras import backend as K
 from keras import layers, models
 from keras.utils import np_utils
-from keras.backend import relu, sigmoid
+from keras.backend import relu, softmax
```
Comment (puneith) [inline census/keras/trainer/model.py]: Good change, thx!

File: census/keras/trainer/model.py [hunk 4]
```diff
@@ -117,21 +120,34 @@ def to_numeric_features(features):
...
     features.pop(col)
 
+  #Re-index dataframe (in case categories list changed from the previous dataset)
+  if feature_cols is not None:
+      features = features.T.reindex(feature_cols).T.fillna(0)
+
   return features
 
-def generator_input(input_file, chunk_size):
+def generator_input(input_file, chunk_size,batch_size=64):
```
Comment (puneith) [inline census/keras/trainer/model.py]: Add space between command and batch_size.

**Comment by eeveeta** [comment]
Thank you for this! I tried the new code and I always get the same prediction array on the online prediction even when I try completely different `sample.json` files, but at least it's not predicting both classes at the same time anymore.

**Comment by RemiTurpaud** [comment]
Hey @eeveeta, what are you trying exactly? Can you share a test case?

File: census/keras/trainer/model.py (hunk 3 above)
Comment (RemiTurpaud) [reply census/keras/trainer/model.py]: @puneith: Do you see something else that should be changed here?

File: census/keras/trainer/model.py (hunk 2 above)
Comment (RemiTurpaud) [reply census/keras/trainer/model.py]: Yes, SGD gets stuck predicting the first class (having a ~75% base rate in the training dataset).
RMSprop (and Adam as far as I have tested), takes off towards .8 accuracy rather fast, with the default setup in this example.

**Comment by puneith** [comment]
@RemiTurpaud Sorry for the delay. I am on it now and will try it now. Can you also share your dataset on which you were getting stuck.

**Comment by RemiTurpaud** [comment]
@puneith I used the training dataset provided in the example  (readme) and the same calls.

**Comment by puneith** [comment]
@RemiTurpaud I was able to confirm the changes and they look good 👍. Thank You!

Can you please ensure when you merge the PR that the code works for Py2 and Py3, since this code was branched before we made those changes.

**Comment by RemiTurpaud** [comment]
@puneith The code didn't work with Py3 (nor does the current master).
Made it compatible for both versions and merged your commit fc1c385.
Tested with both python 2 and 3, cloud and local.

File: census/keras/trainer/model.py [hunk 5]
```diff
@@ -15,23 +15,26 @@
...
-from keras.backend import relu, sigmoid
+from keras.backend import relu, softmax
 
-from urlparse import urlparse
+#Python2/3 compatibility imports
+#   urlparse renamed urllib.parse
+try:
+    from urllib.parse import urlparse
+except ImportError:
+    from urlparse import urlparse
```
Comment (puneith) [inline census/keras/trainer/model.py]: use six

File: census/keras/trainer/model.py [hunk 6]
```diff
@@ -15,23 +15,26 @@
...
 from keras.utils import np_utils
-from keras.backend import relu, sigmoid
+from keras.backend import relu, softmax
 
-from urlparse import urlparse
+#Python2/3 compatibility imports
+#   urlparse renamed urllib.parse
+try:
+    from urllib.parse import urlparse
+except ImportError:
```
Comment (puneith) [inline census/keras/trainer/model.py]: use six

**Comment by puneith** [comment]
Thanks @RemiTurpaud added one comment.

**Comment by RemiTurpaud** [comment]
Thanks for the suggestion @puneith! Done

**Comment by puneith** [comment]
@RemiTurpaud The py2/3 compatibility code needs dependencies which are not in requirements.txt and setup.py. This needs to be fixed.

**Comment by RemiTurpaud** [comment]
Hi @puneith , sorry, was away for a while. Agreed and created PR #164
